loader(7) for aarch64

r1mikey opened the pull request:
**NOTE**: Requires https://github.com/richlowe/arm64-gate/pull/55

Introduce `loader(7)` for aarch64 systems.

Building on top of our global switch to U-Boot, which offer UEFI booting, introduce `loader(7)` to aarch64. The requirements of the loader codebase are already covered by U-Boot, EDK2 and AMI systems, so there is very little to do in terms of portable usage of UEFI boot services.

The PR starts with two small patches (adding new UEFI enum values and adding `btxld` to our build tools for use in loader versioning).

In an attempt to keep `loader.conf` aligned between platforms, we then introduce the `BASEARCH` and `IMPLARCH` variables. Note that there are still bugs in the Forth code that prevent the desired reuse (worked around at image creation time in https://github.com/richlowe/arm64-gate/pull/55).

We then add build support for aarch64 systems before adding aarch64 support to the libraries used in loader. In many cases this is simply Makefile support, but in cases where we import code, the code is imported from FreeBSD (much like the `i86pc` implementation).

We add aarch64 support to the loader itself, which broadly means conditionally removing the 4GiB allocation address limit. This is needed because there is no expectation on aarch64 systems that physical memory will exist at any particular address. We also add the ability to allocate an extra page of memory when loading a file. This is then used to align the `dboot` entry point when booting `unix` (needed for self-relocation).

With the basics in place, we add a small library that abstracts UARTs and implements UART handling for the PL011 UART (found on all of our currently supported hardware). The library also keeps environment handling in one place, as well as implementing output-avoidance when loader is producing serial output for the active UART.

We then add support for `libfdt`, add `libfdt` itself and add a utility library for `libfdt`. The utility library is used to obtain the FDT pointer and to perform fixups on the DTB. There is mapping code in this library to derive the same implementation architecture as is set by `inetboot`. Finally, we add a UART discovery library for FDT, driven by the contents of the `/aliases` node. This allows UART name assignment based on the FDT in a manner compatible with Linux, meeting user expectations.

The next section of the series moves ACPICA headers to common, fixes up (or works around) header clashes, implement ACPICA for loader (this is largely based on code from edk2), then uses ACPICA to implement UART discovery for ACPI systems. There are a few things to be aware of here. The first is that there is what looks suspiciously like a C standard library in loader, but that library is incomplete. There is also the FreeBSD standalone library (libsa), which in many cases conflicts with the other C library declarations. The code has been structured to avoid these conflicts. Upstream FreeBSD has been cleaning this all up, which is work we should look to import in the future. The other thing to be aware of is that determining the UART clock frequency is dependent on the quality of the ACPI implementation. It is common to find a `clock-frequency` property in the `_DSD` properties for an ACPI device, and this is what we support in this iteration. Once we have access the systems that model clocks in the manner described by later ACPI specifications we'll need to revise this code to understand that clock modelling.

In all cases, when we can't derive a clock (either due to firmware deficiencies or due to the UART configuration being locked without information about the clock frequency), we record the frequency as 0, and in some cases the BAUD as 0 (i.e. when the SPCR does not contain this information). The intention is that we then avoid touching BAUD settings when running with this configuration.

With ACPI support in place, we identify Arm systems that have ACPI as having an implementation architecture of `ARMH,sbbr` (Arm Holdings, Server Base Boot Requirements) - this is the implementation architecture we'll later use for ACPI support in `unix`.

Finally, we deliver `loader(7)` for aarch64 systems before moving on to adding support in `unix`.

For `unix` support, we create a self-relocating direct-boot (dboot) stub, meaning `unix` can be loaded to any suitably aligned physical memory location (the `dboot` entrypoint must be aligned to a page boundary). The `boot` stub reuses our previous "[shared boot](https://github.com/richlowe/illumos-gate/pull/129)" code for console interaction and configures itself from the FreeBSD-style loader module and module metadata passed by `loader(7)`. This departs from `i86pc`, which massages that data into `multiboot2` format - we don't use `mb2` on aarch64 because the specification is very `i86pc`-specific.

Configuration of the direct-boot stub then uses the firmware tables (UEFI+FDT or UEFI+ACPI) passed by `loader(7)` to complete the configuration, initialising memory management and page tables, including the upper page tables used by `unix`. We still set up mappings for physical memory in `segkpm` space, but no longer establish high mappings for MMIO (we do, however, establish low mappings for the MMIIO we need during early boot).

The `dboot` stub finishes off by loading the kernel from the memory loaded by `loader(7)` to new (high) mappings, set up at the final kernel execution address. `dboot` then jumps to the kernel code loaded at the final VA.

In `unix`, nothing changes until `consconfig_dacf`. The `consplat` functions have been reimplemented, mostly in-line with `i86pc`, to select the console based on the configuration passed by `loader(7)` rather than reinterpreting the FDT.

With all of that in place, we remove `inetboot` before cleaning up the kernel startup path in both `dboot` (which starts up the boot processor) and `unix` which starts up application processors (secondary CPUs). This change is necessary for booting on newer systems (Neoverse-N2), with VHE. There were a few bugs in the way we booted, and this change both addresses those bugs and makes it possible for us to run on systems with EL2 implemented, with or without VHE. The effect of this change is that when booting on systems with EL2 and VHE, the kernel will run at EL2, configuring the unified EL1/EL2 environment. When booted at EL2 without VHE, the kernel will install a stub trap table at EL2 (for Hypervisor dispatch) before dropping to EL1. When booted at EL1 the kernel will remain at EL1. This approach is consistent with recent FreeBSD changes which pave the way for bhyve/aarch64 support.

And, finally, now that we have `loader(7)` we need to keep it up to date. Implement `installboot`, wire it up the `libbe` and hook up `bootadm`.

* Qemu/virt, booted with the UART as console: https://gist.github.com/r1mikey/812a5ed53835e74346282e22423f202e
* rpi4, framebuffer absent, booted with the UART as console: https://gist.github.com/r1mikey/3b063869f87415e10834bdf26e0d27a0
* rpi4, framebuffer present, booted with the UART as console: https://gist.github.com/r1mikey/9a4b5380a1b87861e26e04ecf36d59f4
* rpi4, framebuffer present, booted with framebuffer as console: https://gist.github.com/r1mikey/0d1bb65dc61d185a3d5ac7e828744399 (**do not do this** - you will lose input unless you already have the network set up)

r1mikey committed:
acpi: Move ACPI headers from intel to common

This is a simple move of the `sys/acpi` subdirectory to a shareable
location, along with the required build infrastructure changes to
support the move.

r1mikey committed:
loader: Add aarch64 build support to libficl

This change only requires Makefiles - the rest of libficl is already
portable.

r1mikey committed:
loader: Add aarch64 support to libsa

Aside from Makefile support this change simply imports code from
FreeBSD.

r1mikey committed:
loader: Add aarch64 support to stdarg.h

r1mikey committed:
loader: Add aarch64 support to common

On aarch64 systems the kernel must be the first loaded module. The kernel is
loaded via `LOAD_RAW`, and when a `LOAD_RAW` is performed and no files are
yet loaded we allocate an extra page for loading the file. This allows
the loader to (later) align the file contents in such a way that the dboot
entry point is on a page boundary, which is required for the self-relocation
code to work.

r1mikey committed:
loader: Add aarch64 support to libefi

Aside from Makefile support, this change tweaks an EFI header to
ensure that int types work for aarch64.

Worth noting is that the time source used (time_event.c) is chosen
because it uses UEFI Boot Services. The implementation in time.c uses
UEFI Runtime Services - we can't use this as we may not be at the final
exception level, and calling Runtime Services from different exception
levels is unsupported.

r1mikey committed:
loader: Set up build environment for aarch64

r1mikey committed:
loader: Add aarch64 arch support for EFI loader

Some of this change are direct imports from FreeBSD (cache.c, cache.h).

The existing FreeBSD linker script had a few flaws in our environment.

`nullconsole.c` and `spinconsole.c` are copies of the code in libi386,
and they're not at all architecture-sensitive.

The `start.S` file includes a handcrafter PE header, which allows
us to produce an EFI application via `objcopy`. Having a working PE
conversion in binutils would be ideal, but this will do for now.

`start.S` was also updated to use pc-relative addressing, ensuring that
all early code is position independent.

The `exec.c` file has support for using the loaded kernel (the whole
file), resolving symbols in that file and simply jumping to the loaded
dboot entry point. Loading of the file uses new new "extra page" load
method, allowing the dboot entrypoint to be page-aligned, which is
a prerequisite of the dboot self-relocation code.

r1mikey committed:
loader: Initial EFI loader support for aarch64

* Add support for the UEFI framebuffer on aarch64.
* Support the LOAD_RAW_PLUS_PAGE memory allocation style, used to load
  the aarch64 kernel as a whole file.
* Remove the 4GiB memory allocation limit for load addresses and for
  the heap. Some aarch64 systems (Qemu sbsa-ref comes to mind) contain
  no DRAM below the 4GiB mark.
* Trim i86pc-specific consoles from the loader console list.

r1mikey committed:
loader: Deliver loader for aarch64

Add `loader(7)` to the aarch64 build and package it up. This change
should only be merged alongside the changes to implement direct-boot
support.

r1mikey committed:
loader: Add MMIO UART scaffolding and pl011 support

The MMIO UART framework abstracts UART implementations from the rest
of `loader(7)`, supporting simplified UART implementations (only
`pl011` is currently implemented) while implementing the expected
`loader(7)` behaviour only once.

This framework contains a set of flags to guide the logic of fetching
and setting UART configuration. There are three types of configuration:
1. Firmware-specified, pre-programmed settings
2. Hardware-specified, pre-programmed settings
3. Software-specified settings

Case 1 deals with settings specified by the FDT on a
`/chosen/stdout-path` node and the ACPI SPCR. These settings should
never be changed (but: see below for an escape-hatch).

Case 2 deals with a lack of settings specified by the FDT on a
`/chosen/stdout-path` node. The settings should be derived from what
the loader (u-boot) has programmed into the hardware.

Case 3 is for regular UARTs, which should be programmed in the usual
way.

In locked-settings cases (1 and 2), the user can set the `ignore-spcr`
environment variable to "true" to override the lock and program the
desired settings. This, however, is discouraged and one should prefer
updating the firmware with the desired settings.

This framework is intended to be driven by firmware-specific
configuration logic, added in subsequent diffs.

r1mikey committed:
loader: Discover consoles from FDT

This change implements UART discovery from FDT. Only `pl011` is
supported at the moment. This library drives the MMIO UART library
for the implementation of the UART drivers.

Discovery is driven by looking through the /aliases node for entries
matching `serial0`-`serial25`. The serial index becomes the `tty` name,
with `serial0` becoming `ttya`, `serial1` becoming `ttyb` etc.

The code uses `/chosen/stdout-path` to determine which UART is the
system stdout. UART configuration is parsed from this `/chosen` node.
If configuration was present, then the UART is set as having
firmware-specified configuration. If no configuration was present, then
the UART is set as having hardware-specified configuration.

Clock frequency is required for setting UART configurations, so the
library implements clock frequency resolution for `fixed-clock` and
BCM2835 CPRMAN clocks (a.k.a Raspberry Pi 4). Frequency resolution
falls back to the `clock-frequency` property on the UART node, which
is common in simple FDT trees.

The BCM2835 CPRMAN implementation uses the BCM2835 MBOX facility, which
also requires dcache invalidation for a range of addresses, so this
diff adds that code.

This diff also includes the required register and ranges resolution
and dma-ranges resolution (both bus-to-cpu and cpu-to-bus), as needed
by the clock resolution code.

Finally, this diff hooks up the FDT UART discovery in the consoles
table.

r1mikey committed:
loader: Add acpica for loader on aarch64

This change provides acpica compiled for the loader environment. There
are three parts to this change:
* Simple Makefile support
* Implementation of mandatory integration functions, based closely on
  those found in edk2, but cleaned up to avoid undersirable behaviour
  and properly handle hardware reduced platforms. This includes removing
  support for legacy ioports, which don't exist on aarch64.
* A simplified set of initialisation and finalisation functions to be
  used around usage of ACPI.

Aside from the init/fini functions, the most interesting part of this
change is that the default integration functions would call UEFI
Runtime Services to get the system time. On aarch64 systems one can't
call UEFI Runtime Services until you are at your final exception level,
and since we don't know what that will be at this stage in the boot, we
have reimplemented the `AcpiOsGetTimer` function in terms of the
architected timer. This function is called as part of avoiding infinite
loops when interpreting AML, so it will always be called during
namespace initialisation and may be called while interrogating the
system bus namespace.

r1mikey committed:
loader: Discover consoles from the ACPI DSDT

Similar to the FDT UART discovery library, this diff adds UART discovery
from the ACPI DSDT.

The DSDT is traversed, looking for compatible UARTs. When a compatible
UART is found, the library walks the resources for the device,
extracting the MMIO regions, then walks the device-specific data,
looking for the properties UUID. If the properties UUID is found, the
library attempts to extract the `clock-frequency` for the device when
one is present.

When a device matches the SPCR, the device attributes are set per the
configuration in the SPCR and, if no `clock-frequency` was found, the
clock frequency is set to the value in the SPCR (if the SPCR is a new
enough version that it contains this field).

Once the system bus has been traversed, the library attempts to assign
device indices to the discovered devices. The most common method of
assigning indices is when the _HID value for all discovered devices is
the same, in which case the ACPU _UID can be used as a source of port
numbering. Failing that, the code attempts to derive nunbering from
the device node names when all name stems are the same. In the future
we might also fall back to _DDN, but that seems like a bit of a hail
Mary. If all index-assignment logic fails the devices are assigned
indexes in discovery order.

Once indexes are assigned the library creates MMIO UART objects for
each discovered device in the usual way, making the UARTs visible as
consoles to loader.

r1mikey committed:
loader: Fix some declaration clashes for aarch64

The loader code contains two competing implementations of C library
functionality (those that live in standard-like headers and those that
live in stand.h). These declarations are often incompatible, which
creates challenges when mixing code that needs one style with code that
needs the other (in this case, acpica and loader code).

Add a few conditional tweaks to avoid the undesirable (and in some cases
unimplemented) parts of the FreeBSD libc that could otherwise be
referenced in a loader build on aarch64.

r1mikey committed:
loader: Add BASEARCH and IMPLARCH variables

Adding `BASEARCH` allows us to replace the parts of the loader
configuration that reference `i86pc` with this variable.

The `IMPLARCH` variable allows us to select a platform (the names are
used interchangeably) to boot and to set up the loader module
(load-file) path portably.

r1mikey committed:
loader: Add an FDT utilities library

The FDT utilities library creates, on-demand, a read/write copy of the
DTB blob provided by UEFI. Prior to providing the blob to a caller,
the devicetree is normalised to include nodes expected by illumos and
all nodes are updated to include a phandle, which is required by the
illumos device tree integration.

The FDT blob is only ever returned when the booted system does not
offer ACPI tables. The Arm EBBR states that when both firmware table
implementations are present that only ACPI should be offered, so this
is just an extra enforcement point that should be unnecessary if the
bootloader is compliant.

The library also includes code for deriving platform details, much
the same way as `inetboot` has done in the past.

Both changes are hooked into loader in this diff.

r1mikey committed:
loader: Add libfdt

r1mikey committed:
unix: Prepend a loader, similar to dboot

The dboot stub for aarch64 systems differs from i86pc dboot in
significant ways. While i86pc boots using multiboot2, mb2 is
unsuitable for non-x86 architectures. On aarch64 systems there are no
expectations of what the memory map will look like, and it is often
the case that there is no DRAM mapped below 4GiB.

For these reasons, the basic idea behind the aarch64 dboot is:
1.  dboot can be loaded anywhere in the first 256TiB of physical address
    space, as long as the dboot entrypoint is on a 4KiB-aligned
    boundary.
2.  dboot is a position-independent standalone executable, and will
    self-relocate prior to running the loader code.
3.  dboot interprets FreeBSD-style modules and module metadata directly,
    rather than having loader translate this metadata to multiboot2.
4.  dboot shares boot code with unix, much like i86pc does.
5.  dboot expects to be loaded by a UEFI loader, which has left the MMU
    enabled and identity-mapped all physical memory. dboot then disables
    the MMU before setting up it's own lower identity mappings and I/O
    mappings.
6.  dboot populates identity mappings for memory into segkpm space, much
    like inetboot used to do, but unlike inetboot it does not map any
    I/O into segkpm. This means that the device arena must be used for
    I/O in unix (as one would expect). These mappings happen in upper
    space (TTBR1, the top 265TiB of VA).
7.  dboot loads the kernel into it's final VA location and sets up
    appropriate mappings prior to jumping to the kernel. This is
    particularly important, as while dboot executes from within the
    memory that was loaded with the entire unix file it copies the
    loadable program segments into new memory before jumping to them,
    which means that one cannot reason about relative locations to find
    the dboot code and data from unix.
8.  dboot papers over differences in firmwares that would affect
    early boot. At the moment this means architected timer frequency
    discovery and PSCI, but this will be extended in the future to also
    cover cpuinfo.
9.  dboot passes memlists to unix, abstracting firmware details. This is
    similar to i86pc, where the e820 map is abstracted, but dboot drives
    memory discovery from the UEFI memory map.
10. dboot does not include a huge scratch area like inetboot did. The
    scratch heap is allocated and passed by loader.

Basic flow is:
* Relocate
* Initialise early console (if applicable)
* Discover configuration
  * Discover configuration from FreeBSD-style modules. This gets
    unix and related configration metadata, the root filesystem and
    the loader font (if present).
  * As console information is discovered the console is activated.
  * When the environment is discovered (this is metadata on the
    kernel) it is interrogated to determine if dboot debugging should
    be enabled.
  * Boot flags are massaged, which can also enable verbose output
    from dboot (using -V).
  * Discover configuration from the firmware tables:
    * On FDT systems this means retrieving PSCI information,
      initialising PSCI, then getting the architected timer frequency
      when specified in firmware tables.
    * On ACPI systems this means getting PSCI configuration from the
      FADT table and initialising PSCI. ACPI systems do not provide
      information about the architected timer frequency, which must
      have been programmed into the hardware from an earlier boot
      loader.
  * Retrieve the architected timer frequency from the hardware if not
    already discovered.
  * Process the retrieved physical memory map, creating memlists,
    filling the page tables and turning on the MMU.
* Load the ELF payload (unix)
* Jump to the loaded payload

Worth noting is the dboot performs CPU initialisation for the boot CPU
only, while the secondary CPU startup code in unix initialises
application CPUs. This is a slightly unfortunate split of code, but is
effectively the same as what we've had to date. This is particularly
relevant in a subsequent diff, which handles running unix at EL2.

r1mikey committed:
psm: Remove aarch64 inetboot

Now that we have a direct-boot kernel for aarch64 we no longer need
the `inetboot` boot shim.

r1mikey committed:
boot: Allow the kernel to run at EL2

At the moment we always drop to EL1 to run the kernel, completely
ignoring any Hypervisor functionality offered by the CPU. Given that
we don't yet support Bhyve we have no need for EL2 at present. However,
our boot process does not survive booting on systems that offer
Virtualisation Host Extensions (VHE). There are many reasons for this,
including sign-extension bugs and the fact the some register bits are
different when VHE is present. This was uncovered when booting on the
Qemu SBSA reference VM, which offers a Neoverse-N2 CPU by default.

On real hardware, the Raspberry Pi 4 offers EL2, but without VHE, as
does Qemu virt with the `virtualization=on` machine flag. In these
cases we need to install a virtualisation trap handler to EL2 prior
to dropping to EL1. We install a stub trap table (much like FreeBSD),
allowing us to replace the table with a full trap table later. Ampere
Altra and Altra Max systems are similar to Pi4 in this regard.

On modern hardware with VHE (Qemu sbsa-ref, AmpereOne, Grace) we need
to run the kernel at EL2, reducing the number of context switches
needed for Hypervisor operations. In this setup we do not install a
stub trap table, as we run at EL2 (so we can install a table at any
time).

So, we have:
* Boot at EL1, no Hypervisor extensions.
* Boot at EL2, no VHE, install stubs, drop to EL1.
* Boot at EL2, VHE, configure unified EL12 and stay at EL2.

Worth noting is that due to the way we boot, we replace the stub
trap table while starting the kernel. This is because the code
executed by `loader(7)` is backed by different physical pages than
the running kernel and because booting secondary processors uses
temporary boot pages.

The side-effect of this change is that we can boot to the end of
the loader on Qemu SBSA, which implements Neoverse-N2 with VHE. This
has also been tested to boot into the kernel (but only the early
kernel, as ACPI support is still a work in progress for the kernel).

The revised startup code has heavy reference to FreeBSD, with the
intent of adding Bhyve/aarch64 support in the future (once that is
fully merged into FreeBSD and stable).

r1mikey committed:
tools: Add aarch64 btxld to prepare for loader

The `btxld` utility is used to insert the loader version into an ELF
binary at at build time, facilitating later conditional updates
via `installboot` and `bootadm`.

r1mikey committed:
boot: Add installboot for aarch64

Now that we have `loader(7)` for aarch64 we can start plumbing it in
to the system administration flow so that image updates work as
expected (i.e. appear to work like they do on i86pc).

Adding `installboot` support is the first step to supporting both
`bootadm` and `beadm`.

r1mikey committed:
boot: Teach libbe about the aarch64 installboot

Now that we have `installboot` support on aarch64 we can plumb that
support into `libbe`, laying the foundation for properly supporting
boot environment management.

r1mikey committed:
boot: bootadm should support menu on aarch64

As the last step in enabling boot environment management, add support
for the boot volume `menu.lst`. While this seems `grub`-specific, this
file is used by `loader(7)` to list installed boot environments.

r1mikey committed:
loader: Add the ARMH,sbbr implementation architecture

ACPI-based aarch64 systems will broadly comply with the Arm
SystemReady-SR or Systemready-ES standards families (and might comply
with the SystemReady-IL standards).

Since illumos ACPI support is intended to be a fairly permissive
interpretation of the Server Base Boot Requirements (DEN0044 [1]),
the implementation architecture (platform) will be named `ARMH,sbbr`.

There's no kernel for `ARMH,sbbr` at the moment, but this seems like
an opportune time to recognise what it will be and add support to the
loader.

[1] https://developer.arm.com/documentation/den0044/latest/

r1mikey committed:
dacf: Flesh out consplat for aarch64

Flesh out the platform console selection for armv8/aarch64 platforms.

This change merges in logic from the `i86pc` implementation, so
copyrights are updated accordingly.

Alongside those changes, we recognise `ttya`-`ttyz`, as there's no
architected notion of built-in consoles as one might find on `i86pc`.

The text (framebuffer) console is recognised, as are keyboards and mice,
none of which are in place yet, but will be put up for review shortly.

This is necessary so that we respect the console settings provided by
loader, rather than looking at the FDT.

r1mikey commented:
@hadfl, @richlowe - this is the remainder of the loader changes, along with the required arm64-gate build changes needed to get things up and running.

EFI text output (used when no framebuffer is connected) is a bit slow under U-Boot, but it's all otherwise usable.

The full stack includes installboot and friends, which should make future updates easier.

I'm putting this up in draft for now, and will rebase it once https://github.com/richlowe/illumos-gate/pull/143 has landed. I don't foresee more structural changes, so any review comments would be appreciated.

I'll continue uploading testing results for other systems (Qemu Virt with OVMF, Qemu SBSA-ref with EDK2, Raspberry Pi 4 with EDK2 and Altra Max with AMI BIOS), but all of those only exercise `loader(7)` for now, not `unix`.

richlowe commented:
This is so cool.  I have booted this both as it is, and merged with my PCI stuff.  It seems great.

Would you be ok with me taking changesets out of this branch that are in preparation for the big changes?  Things like building libefi that can be cherry-picked easily and are self contained?

That might help with the biggest problem I have -- github chokes when I attempt to review it here :)

r1mikey commented:
> Would you be ok with me taking changesets out of this branch that are in preparation for the big changes? Things like building libefi that can be cherry-picked easily and are self contained?

Absolutely, yes.

Everything up to "loader: Deliver loader for aarch64" (which hooks loader up to the build and packaging) can be individually merged.

With that in, the next five changes simply prepare for dboot (the next big one), and can be individually merged. Once dboot is in the remainder are safe to individually merge as well.

I'm just re-validating the edk2 and AMI platforms, then I'll rebase this and publish it.

richlowe commented:
So unless someone objects, I think it would be easier to take some of the lower preparatory pieces out of this branch like I said, which I can do.  

The other thing we could do (and I have done before) is use the illumos gerrit, which handles changes of this size better.  If anyone wants that, ping me and I can work it out.

My only review comment so far (before my browser tab crashes), is that you should not need to build in src/stand or src/psm on aarch64 anymore.  Since SPARC is obsolete, you could even delete them if you wanted to

r1mikey commented:
> My only review comment so far (before my browser tab crashes), is that you should not need to build in src/stand or src/psm on aarch64 anymore. Since SPARC is obsolete, you could even delete them if you wanted to

We're still using memlists and standalloc from psm at this point. This is getting in my way elsewhere, so I'll get that cleaned up in a follow-up change.

I can split this up into a series of PRs for review if that would help?

richlowe commented:
I was about to say, maybe a series of PRs is better than me trying to take things because me doing it would make an unimaginable merge hell for you.  Yes, that would be great.  Thanks.

What I had done in my branch here is take it through "loader: Initial EFI loader support for aarch64" for a first look at things.

tsoome commented:
Oh, sorry, I forgot the github review is not usable with multiple commits.

tsoome commented on usr/src/tools/Makefile:
In my tree, I have:
tsoome@beastie:~/illumos-gate$ git diff 1b800aa8dcb5492ab2d5ac09c602e8c65906d366..190366545e17e4570272b39c1f5598539fb7012b
diff --git a/usr/src/tools/btxld/Makefile b/usr/src/tools/btxld/Makefile
index 599cf5f406..bdec8e7671 100644
--- a/usr/src/tools/btxld/Makefile
+++ b/usr/src/tools/btxld/Makefile
@@ -27,8 +27,8 @@ PROG=         btxld
 MAN1ONBLDFILES=        btxld.1onbld
 SRCS=          btxld.c elfh.c version.c $(UTILS_SRC) $(EINFO_SRC) $(EXTRA_SRC)
 OBJS=          btxld.o elfh.o version.o bblk_einfo.o mboot_extra.o boot_utils.o
-LDLIBS +=      -lmd5
-NATIVE_LIBS += libmd5.so
+LDLIBS +=      -lmd
+NATIVE_LIBS += libmd.so
 CSTD=  $(CSTD_GNU99)
 CPPFLAGS +=    -I$(SRC)/uts/common -I$(COMMON_SRC)
 
tsoome@beastie:~/illumos-gate$

that is, for some reason there was complaint about linking against libmd5. I did it 2023, so I'm not too sure about the exact details. Need to investigate...

r1mikey replied:
I see the upstream review is up at https://code.illumos.org/c/illumos-gate/+/4192 - well pick this up (and `:$s/NATIVE_LIBS/ADJUNCT_LIBS/g`) when we next merge onto upstream (cc: @richlowe )

tsoome commented on usr/src/uts/common/sys/efi.h:
This needs to land on gate.

r1mikey replied:
I think we can pull it up to gate later? If not I'll just remove the handling of that memory type (it's pretty new, and I have yet to see a machine that uses it).

tsoome replied:
I was actually thinking to pick up your commit - it is nice self containing onliner and backed by UEFI specification, so it should be easy....

r1mikey replied:
I've split this commit out for now and will send it in via gerrit. Code using this value has been conditionally disabled for now.

tsoome commented on usr/src/boot/Makefile.inc:
```diff
@@ -42,20 +42,30 @@ LZ4=		$(SRC)/common/lz4
 AS_CPPFLAGS=
-CPPFLAGS=	-D_STANDALONE -_gcc=-nostdinc
-CFLAGS64=	-_gcc=-mno-red-zone
+CPPFLAGS=	-D_LOADER -D_STANDALONE -_gcc=-nostdinc
```
Why _LOADER? the _STANDALONE is currently used to have bootloader specific code sections in current gate and also in OpenZFS code tree (for FreeBSD loader).

r1mikey replied:
_STANDALONE conflicts with considerations made for the illumos `libsa`, which is not entirely compatible with the `libstand` in the code imported from FreeBSD.

illumos `libsa` is currently used in the dboot implementation for aarch64, so there is a need to disambiguate.

tsoome replied:
I see. Well, I think we can come back to it sometimes later;)

r1mikey replied:
There is follow-up work to stop using anything in `usr/src/psm` and `usr/src/stand` (moving standalloc and other required code into the dboot implementation), at which point this can be cleaned up (and the SPARC-specific code deleted).

r1mikey replied:
This is removed in the latest push, as we no longer reference the illumos `stand` and `psm` code.

tsoome commented on usr/src/boot/libficl/Makefile:
```diff
@@ -15,7 +15,9 @@
 
-SUBDIRS = softcore $(MACH) $(MACH64)
+i386_SUBDIRS = $(MACH) $(MACH64)
+aarch64_SUBDIRS = arm64
```
Is there some document describing the MACH and MACH64 choices and such? subdirs should use $(MACH64) here, isn't it so?

r1mikey replied:
Only IPD24, which only alludes to this.

On illumos/aarch64, both MACH and MACH64 are set to `aarch64`, but the FreeBSD tree we import from uses `arm64`. The intent here is to keep things looking like FreeBSD to make future imports easier.

tsoome replied:
Yea, I do know that feeling. Well, there are basically two options -- either we keep fbsd way, but then we really should set MACH64 at  usr/src/boot master makefile and then we do not have to duplicate it every significant place.

Other option is to use aarch64, have a bit of annoying work with pulling few bits from fbsd but we are following the logic of the rest of the system. Quite frankly, we are not pulling that much/often from FreeBSD, so this is not too big of the issue IMO. In my local experiment I did went for aarch64, as it felt more logical choice at that time (2023;), but I do not mind either way.

Anyhow, it is just about the agreement and decision -- only thing is, we should fix it now rather than later;)

r1mikey replied:
I've move this into the makefiles in `usr/src/boot`.

I dream of the day when BIOS boot and 32-bit UEFI will go away - this would all be so much cleaner without them.

r1mikey replied:
This has been moved in the latest pushes. I introduced `usr/src/boot/Makefile.master` as a place to hold the change, and included that everywhere `usr/src/Makefile.master` was being included.

`Makefile.inc` was unsuitable as it's not included by top-level Makefiles.

tsoome commented on usr/src/boot/libficl/arm64/Makefile:
```diff
@@ -0,0 +1,31 @@
+
+include $(SRC)/Makefile.master
+
+MACHINE=	arm64
```
MACHINE=        $(MACH64)

and missing

$(PICS): machine

CLOBBER += machine

before FRC:  -- we only need PICS as we only do build shared objects to construct efi app.

r1mikey replied:
This is the same as `usr/src/boot/libficl/amd64/Makefile` at the moment (modulo removing an amd64 compiler flag) - is it wrong in both places?

tsoome replied:
the fbsd logic about header is that we create machine and platform specific links for respective headers; if we are building for the platform, then those links need to be created first and thats why we do need the dependencies.  And cleanup as well:) The side effect of such arrangement is number of stub header files redirecting to machine specific headers:)

richlowe replied:
`$(MACH64)` is not "arm64", so probably not that bit toomas?

tsoome commented on usr/src/boot/libsa/Makefile:
```diff
@@ -15,7 +15,9 @@
 
-SUBDIRS = $(MACH) $(MACH64)
+i386_SUBDIRS = $(MACH) $(MACH64)
+aarch64_SUBDIRS = arm64
```
same MACH64 here

tsoome commented on usr/src/boot/libsa/arm64/Makefile:
```diff
@@ -0,0 +1,40 @@
+
+include $(SRC)/Makefile.master
+
+MACHINE=	arm64
```
ditto

tsoome commented on usr/src/boot/libsa/arm64/Makefile:
```diff
@@ -0,0 +1,40 @@
+	$(COMPILE.S) -o $@ $<
+
+include $(BOOTSRC)/Makefile.lib
+
```
I have dependency in my version:
$(OBJS) $(PICS): machine

but we should not need OBJS here, so just PICS should do

r1mikey replied:
This is also directly based on the amd64 code in gate, which does not have that. Are you planning to integrate any of these changes soon?

tsoome replied:
with gate, the "soon" is not defined. I'd like to obviously, but it is not up to me. I know, this sucks, but here we are. (I'm currently trying to get 5 years old commits integrated;)

tsoome commented on usr/src/boot/include/stdarg.h:
```diff
@@ -3,4 +3,8 @@
 /* $FreeBSD$ */
 
+#if defined(__aarch64__)
+#include <machine/stdarg.h>
```
it only needs
#include <machine/stdarg.h>

and need x86/include/stdarg.h. I have the header fix in my update queue, I'll see if i can get to at least post it in gerrit, hopefully get it integrated...

tsoome commented on usr/src/boot/common/module.c:
```diff
@@ -337,8 +337,28 @@ file_load(char *filename, vm_offset_t dest, struct preloaded_file **result)
+	if (preloaded_files == NULL) {
+		if (archsw.arch_loadaddr != NULL)
+			dest = archsw.arch_loadaddr(
+			    LOAD_RAW_PLUS_PAGE, filename, dest);
```
am, we do not need new macro for that, just add aarch64 specific increment in arch_loadaddr implementation. In fact, I would not really repeat the mistake of having dboot at all, but of course, if it is the currently available implementation and does enable the kernel loading and start, then we can fix it later.

The problem about dboot is that it adds one more "kernel" to take care of, and IMO, the functionality in it should be either in bootloader or in locore (early kernel). Ideally, the parts which are only needed to prepare the kernel, should be in boot loader.

dboot originally was introduced because multiboot1 protocol was only dealing with 32-bit machine and so they needed something to collect boot info and switch to 64-bit mode and to drop all this "run once" code as soon as trampoline to locore is taken. Later, with multiboot2, we dod not had much chance than extend dboot to support multiboot2, because it was fastest option to get things done and it was easier to add mb2 support for things like ipxe than to teach them to deal with illumos kernel - although, that is probably debatable, because as always, things do turn out to be more complicate anyhow;)

So, I'd keep things simple here, and would deal with necessities in arch_loadaddr() implementation.

r1mikey replied:
`arch_loadaddr` is used extensively though, so we can't have it adding an extra page everywhere, which is all it could do without additional guidance... or am I missing something here?

The extra page is needed so that we can align the dboot entry point to a page boundary (i.e. we load the raw file and jump to that address, which must be page-aligned for self-relocation to work).

tsoome replied:
no-no, the loadaddr can recognize if we are called for kernel or modules, this way we do not have to add page for other cases.

r1mikey replied:
Oh, you mean to handle LOAD_ELF specifically here? At the moment that just punts:
```
        if (type == LOAD_ELF)
                return (0);     /* not supported */
```

It's also documented as:
```
#define LOAD_ELF        1       /* data points to the ELF header. */
```

So on aarch64 you'd want this to pass data as the file (not the ELF header), stat that, add a page and otherwise treat it like LOAD_MEM?

tsoome replied:
From boot loader point of view, we have two options. Current loader does use MB2 protocol, that is, we read in whole kernel as raw data, we seek multiboot header structure from it to get list of structures it is requesting boot loader to provide and load and entry address. on exec, the raw kernel is copied to load address, mb2 info and modules after kernel,  and we jump to entry address (which is dboot). It does not have to be dboot, it could be locore (text segment from ELF), but that really does not change things much, we can now jump directly from 64-bit loader to 64-bit kernel, because MB2 protocol got some support for loading and running 64-bit kernel.

The alternate way is to perform ELF load (that is, set up text, data, bss), pass needed information (currently xboot_info structure) + modules and jump to address based on ELF header information. This is what you can see happening in FreeBSD loader, and both ELF load and multiboot load you can see in grub2. Doing ELF load implies that we interpret ELF file and have way to pass the machine specific data. Loader can already interpret ELF files, it only needs small adjustments to handle illumos kernel there;)

What we want to achieve is sort of balance; we do not really want to keep too much "run once" code in kernel, because it is waste of memory, and we do not want to overcomplicate boot loader either. But having code in boot loader (and in dboot) does allow us to free that memory once kernel is running.

Using ELF loading does free us from limits of MB2 protocol, but also does complicate support for other boot loaders (where such support is important).

r1mikey replied:
Resolved in the latest push. The first `LOAD_RAW` on aarch64 will add the extra page.

tsoome commented on usr/src/boot/common/module.c:
```diff
@@ -753,6 +773,15 @@ mod_load(char *modname, struct mod_depend *verinfo, int argc, char *argv[])
 	char *filename;
 
+	/*
+	 * On aarch64 systems the first loaded module must be the kernel.
```
this is true on all cases actually. on x86, the check happens in file_loadraw() because in current implementation we do not load ELF, but raw file as is and dboot will copy text, data and bss in place. I have WIP to implement ELF loading instead, to be able to skip dboot at all (so we would not need to do 64-bit -> 32-bit -> 64-bit dance), and dboot would be there for some time to support other boot loaders.

r1mikey replied:
The way aarch64 boots (inherited from the inetboot implementation) is that segkpm VA is offset-mapped into high memory by the loader, which is both truly awful and quite expedient (there's no keyhole surgery to set up the memory map).

On aarch64 we also boot in high memory (set up by dboot), which is a separate translation-table base - this is set up by dboot.

Another complication I can think of is that we're using dboot (and inetboot before it) to iron out a number of platform and firmware specifics before passing control to unix. This includes things like the architected timer frequency, dealing with the UEFI memory map, establishing MMIO mappings for the UART and framebuffer (when present), figuring out how to speak to PSCI and (in the near future) portably communicating the CPU lists to unix. It's possible to pass this to union directly, but at the moment it's trivial to update the xboot_iinfo (it's always going to be consistently defined between unix and dboot). Is this something that concerns you at all?

tsoome replied:
mm, ok, perhaps this is all right. I can not say now, I really haven't read all those its, just having experience from current x86 solution, where we have essentially 4 kernels - boot loader, then dboot, then locore (early kernel) and then final kernel. All with their implementations for communications etc. But as I wrote, if this approach does enable the things to work, it is good and can be improved later (if it does need it).

tsoome commented on usr/src/boot/efi/libefi/efi_console.c:
```diff
@@ -457,8 +457,13 @@ efi_framebuffer_setup(void)
 	}
 	shadow_sz = EFI_SIZE_TO_PAGES(efifb.fb_width * efifb.fb_height *
 	    sizeof (*shadow_fb));
+#if defined(__aarch64__)
```
This needs a comment. on x86, we allocate below max address to avoid memory map fragmentation, because we do need space for loading kernel and modules and so far, the usable space has been limited to 4GB line (because historically there have been machines with memory mapped only up to 4GB). Today, there are cases where it would be helpful to be able to use more memory - mostly because of memory fragmentation and the need to use huge "miniroot" boot archives. But, while we only can use dboot to start the kernel, we are still bound to 4GB limit because dboot is running in 32-bit protected mode.

r1mikey replied:
Yeah, this is all processor-specific in the UEFI spec. The x64 portion (UEFI 2.11, §2.3.4) implies that all memory presented in the UEFI memory map must be identity mapped. I guess this is what happens when there's no certification process in place.

aarch64 (UEFI2.11, §2.3.6) says effectively the same thing, but with clearer wording - and there's a compliance suite that a vendor must pass, which helps to keep them honest. The only vagueness is when PA size is greater than VA size, but that's extremely unlikely.

tsoome commented on usr/src/lib/fm/topo/modules/common/usb/Makefile:
```diff
@@ -23,6 +23,6 @@ SHAREDSRCS	= topo_port.c
 include ../../Makefile.plugin
 
-CPPFLAGS	+= -I../shared  -I$(SRC)/uts/intel/sys/acpi -DACPI_APPLICATION
+CPPFLAGS	+= -I../shared  -I$(SRC)/uts/common/sys/acpi -DACPI_APPLICATION
```
This is the reason, why acpica should be in usr/src/common instead -- that is, those bits are not only needed by kernel, boot loader but also by userland.

r1mikey replied:
Yeah, I've tried to not boil the ocean too much, but `usr/src/contrib` feels like a much better home for the sources and headers.

tsoome replied:
Better to have it boiled in one go;)

tsoome reviewed (commented):
I have left bunch of comments. There are many things I'm not really sure about (because I do not know about the hw specific details and quirks and whatnot). But in general, if it helps to enable things, we have always time for improvements and changes (I do not believe in commits which are carved in stone).

r1mikey commented:
I'm going to split this up into what I plan to be 18 smaller PRs, grouped by size/impact and ordered by dependency. I'll post the list here shortly.

The PRs will stack, so as they are merged I'll rebase the remaining ones and publish the next one for merging.

r1mikey commented:
> should not need to build in src/stand or src/psm on aarch64 anymore.

I'm working on this at the moment - I have it working, but it needs some cleanup. I'll split things up once this is in good shape.

r1mikey committed:
sys/types.h: expose limits to aarch64 dboot

r1mikey committed:
libfdt: Add _BOOT and _STANDALONE support to libfdt_env.h

r1mikey committed:
acpi: Add aarch64 support to ACPI headers

r1mikey committed:
util: Add _BOOT support to strtoul

r1mikey committed:
util: Add _BOOT support to memmove

r1mikey committed:
util: Add _BOOT support to memcmp

r1mikey committed:
util: Add _BOOT support to memchr

r1mikey committed:
util: Add _BOOT support to bzero

r1mikey commented:
I'm just rebuilding from scratch having removed any references to psm and stand - will start splitting things up when that's done and tested.

r1mikey committed:
build: No need to build psm and stand any more

Since we have deleted `inetboot` we no longer need this code.

r1mikey commented:
> you should not need to build in src/stand or src/psm on aarch64 anymore. Since SPARC is obsolete, you could even delete them if you wanted to

This is done in the latest push. I have removed `psm` and `stand` from the subdirs list for aarch64 in the top-level Makefile and cleaned up the exceptions list.

r1mikey commented:
I've ended up with 16 PRs, starting at https://github.com/richlowe/illumos-gate/pull/146.

Each of these will need to be rebased once the preceding one has landed, otherwise they just get progressively larger. One day, in a happy future, Github will support stacked changes.

r1mikey commented:
Closing this as the split stack of PRs is most of the way to being merged.
